SAF: Crash on selecting "Downloads" when granting SAF access

Taken from https://github.com/cgeo/cgeo/issues/9868#issuecomment-770183465 by @baiti:

For completeness, here is the FATAL ABORT for the Grant, do nothin scenario
`2021-01-30 10:25:05.079 12279-12279/cgeo.geocaching D/AndroidRuntime: Shutting down VM
2021-01-30 10:25:05.080 12279-12279/cgeo.geocaching E/AndroidRuntime: FATAL EXCEPTION: main
    Process: cgeo.geocaching, PID: 12279
    java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=59372, result=-1, data=Intent { dat=content://com.android.providers.downloads.documents/tree/downloads flg=0xc3 }} to activity {cgeo.geocaching/cgeo.geocaching.MainActivity}: java.lang.IllegalArgumentException: mimeType can't be null
        at android.app.ActivityThread.deliverResults(ActivityThread.java:4845)
        at android.app.ActivityThread.handleSendResult(ActivityThread.java:4886)
        at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:51)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
     Caused by: java.lang.IllegalArgumentException: mimeType can't be null
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:170)
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:140)
        at android.content.ContentProviderProxy.call(ContentProviderNative.java:658)
        at android.content.ContentResolver.call(ContentResolver.java:2042)
        at android.provider.DocumentsContract.createDocument(DocumentsContract.java:1327)
        at cgeo.geocaching.storage.DocumentContentAccessor.createInternal(DocumentContentAccessor.java:111)
        at cgeo.geocaching.storage.DocumentContentAccessor.create(DocumentContentAccessor.java:98)
        at cgeo.geocaching.storage.ContentStorage.create(ContentStorage.java:169)
        at cgeo.geocaching.storage.ContentStorage.tryTestReadWriteToFolder(ContentStorage.java:426)
        at cgeo.geocaching.storage.ContentStorage.ensureFolder(ContentStorage.java:133)
        at cgeo.geocaching.storage.ContentStorageActivityHelper.handleResultPersistableFolderSelection(ContentStorageActivityHelper.java:338)
        at cgeo.geocaching.storage.ContentStorageActivityHelper.onActivityResult(ContentStorageActivityHelper.java:210)
        at cgeo.geocaching.activity.AbstractActivity.onActivityResult(AbstractActivity.java:300)
        at cgeo.geocaching.MainActivity.onActivityResult(MainActivity.java:571)
        at android.app.Activity.dispatchActivityResult(Activity.java:8110)
        at android.app.ActivityThread.deliverResults(ActivityThread.java:4838)
        at android.app.ActivityThread.handleSendResult(ActivityThread.java:4886) 
        at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:51) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016) 
        at android.os.Handler.dispatchMessage(Handler.java:107) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.app.ActivityThread.main(ActivityThread.java:7356) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930) 
`

--- a/main/src/cgeo/geocaching/maps/mapsforge/MapsforgeMapProvider.java
+++ b/main/src/cgeo/geocaching/maps/mapsforge/MapsforgeMapProvider.java
@@ -355,7 +355,7 @@ public final class MapsforgeMapProvider extends AbstractMapProvider {
         final Resources resources = CgeoApplication.getInstance().getResources();
         final List<ImmutablePair<String, Uri>> offlineMaps =
             CollectionStream.of(getOfflineMaps())
-                .filter(fi -> !fi.isDirectory && !fi.name.toLowerCase().endsWith(OfflineMapUtils.INFOFILE_SUFFIX) && isValidMapFile(fi.uri))
+                .filter(fi -> !fi.isDirectory && fi.name.toLowerCase().endsWith(OfflineMapUtils.MAPFILE_SUFFIX) && isValidMapFile(fi.uri))
                 .map(fi -> new ImmutablePair<>(fi.name, fi.uri)).toList();
         Collections.sort(offlineMaps, (o1, o2) -> TextUtils.COLLATOR.compare(o1.left, o2.left));
         if (offlineMaps.size() > 1) {

--- a/main/src/cgeo/geocaching/storage/ContentStorage.java
+++ b/main/src/cgeo/geocaching/storage/ContentStorage.java
@@ -305,6 +305,10 @@ public class ContentStorage {
 
     /** Opens an Uri for reading. Remember to close stream after usage! Returns null if Uri can't be opened for reading. */
     public InputStream openForRead(final Uri uri) {
+        return openForRead(uri, false);
+    }
+
+    public InputStream openForRead(final Uri uri, final boolean suppressWarningOnFailure) {
         if (uri == null) {
             return null;
         }
@@ -312,7 +316,9 @@ public class ContentStorage {
         try {
             return this.context.getContentResolver().openInputStream(uri);
         } catch (IOException | SecurityException se) {
-            reportProblem(R.string.contentstorage_err_read_failed, se, uri);
+            if (!suppressWarningOnFailure) {
+                reportProblem(R.string.contentstorage_err_read_failed, se, uri);
+            }
         }
         return null;
     }
@@ -421,22 +427,28 @@ public class ContentStorage {
     /** Tries to read and (optionally) write something to the given folder location, returns whether this was successful or not */
     private boolean tryTestReadWriteToFolder(final Folder folder, final boolean testWrite) {
 
-        Uri testDoc = null;
-        if (testWrite) {
-            testDoc = create(folder, FileNameCreator.DEFAULT, false);
-            if (testDoc == null) {
+        try {
+            Uri testDoc = null;
+            if (testWrite) {
+                testDoc = create(folder, FileNameCreator.DEFAULT, false);
+                if (testDoc == null) {
+                    return false;
+                }
+            }
+
+            //actually, on folder without write permission, we can not do much. Simply list the content...
+            final List<FileInformation> files = list(folder);
+
+            if (testWrite && files.size() < 1) {
                 return false;
             }
-        }
 
-        //actually, on folder without write permission, we can not do much. Simply list the content...
-        final List<FileInformation> files = list(folder);
-
-        if (testWrite && files.size() < 1) {
+            return testDoc == null || delete(testDoc);
+        } catch (RuntimeException re) {
+            //if something unexpected happens, then we consider this folder non-readable/writeable
+            Log.e("Error on testing read/write of folder " + folder + ", testWrite=" + testWrite, re);
             return false;
         }
-
-        return testDoc == null || delete(testDoc);
     }
 
     /** Gets this folder's current location. Tries to adjust folder location if no permission given. May return null if no permission found */

--- a/main/src/cgeo/geocaching/storage/DocumentContentAccessor.java
+++ b/main/src/cgeo/geocaching/storage/DocumentContentAccessor.java
@@ -12,7 +12,9 @@ import android.content.Intent;
 import android.content.UriPermission;
 import android.database.Cursor;
 import android.net.Uri;
+import android.os.Build;
 import android.provider.DocumentsContract;
+import android.webkit.MimeTypeMap;
 
 import androidx.annotation.NonNull;
 
@@ -27,6 +29,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.Set;
 
+import org.apache.commons.lang3.StringUtils;
 import static org.apache.commons.io.IOUtils.closeQuietly;
 
 /**
@@ -106,11 +109,59 @@ class DocumentContentAccessor extends AbstractContentAccessor {
         }
         final String docName = createUniqueFilename(name, queryDir(folderUri, new String[]{ DocumentsContract.Document.COLUMN_DISPLAY_NAME }, c -> c.getString(0)));
         try (ContextLogger cLog = new ContextLogger("DocumentFolderAccessor.create %s: %s", folder, name)) {
-            //Do NOT pass a mimeType. It will then be selected based on the file suffix.
-            //Use an empty string for mimeType; null triggers an exception in API21. Tested also in API23, API29, API30
-            return DocumentsContract.createDocument(getContext().getContentResolver(), folderUri, "", docName);
+            return DocumentsContract.createDocument(getContext().getContentResolver(), folderUri, guessMimeTypeFor(docName), docName);
         }
     }
+
+    /** method tested on SDK21, 23, 29 and 30; both with targetSDK=29 and targetSDk=30 */
+    private String guessMimeTypeFor(final String filename) {
+
+        if (StringUtils.isBlank(filename)) {
+            return "";
+        }
+
+        //Note that when passing a wrong mimeType (not fitting to filename suffix),
+        //then most document providers will append the mimeType's default extension
+        //we want to avoid this from happening
+
+        //Android SDK21/22 replace file suffixes with their defaults from mimeType.
+        //This leads to e.g. "image.jpg" to be renamed to "image.jpeg".
+        //->prevent this by using octet-stream for all files
+        if (android.os.Build.VERSION.SDK_INT <= Build.VERSION_CODES.LOLLIPOP_MR1) {
+            return "application/octet-stream";
+        }
+
+        //try guess mimeType from filename suffix;
+        String suffix = filename;
+        final int idx = filename.lastIndexOf(".");
+        if (idx >= 0) {
+            suffix = suffix.substring(idx + 1);
+        }
+        final String mime = MimeTypeMap.getSingleton().getMimeTypeFromExtension(suffix);
+        if (mime != null) {
+            return mime;
+        }
+
+        //special handling of c:geo special files
+        if ("map".equals(suffix)) {
+            return "application/octet-stream";
+        }
+        if ("gpx".equals(suffix)) {
+            //you might be thinking that here "application/xml" or "text/xml" should be returned
+            //but this is not advisable since it would lead Android to append ".xml" to the filename
+            //(would result in filenames like "export.gpx.xml")
+            return "application/octet-stream";
+        }
+
+
+        //as a last resort, return empty string. This seems to work for most of the document providers and API levels
+        //(tested for standard doc provider in APi levels 21,23, 29 and 30. Null triggers an exception in API21)
+        //Noteable exception: "" will trigger exception with "Download" folder provider when the "Download" folder itself is selected.
+        //In this case, ONLY real mime type will help unfortunately. See issue #9903
+        return "";
+    }
+
+
     @Override
     public ContentStorage.FileInformation getFileInfo(@NonNull final Folder folder, final String name) throws IOException {
 

--- a/main/src/cgeo/geocaching/utils/FileNameCreator.java
+++ b/main/src/cgeo/geocaching/utils/FileNameCreator.java
@@ -10,7 +10,7 @@ import java.util.concurrent.atomic.AtomicInteger;
  * */
 public class FileNameCreator {
 
-    public static final FileNameCreator DEFAULT = new FileNameCreator("file", "dat");
+    public static final FileNameCreator DEFAULT = new FileNameCreator("file", "txt");
 //    public static final FileNameCreator DEFAULT_TEXT = new FileNameCreator("file", "txt");
 //    public static final FileNameCreator DEFAULT_BINARY = new FileNameCreator("file", "bin");
 

--- a/main/src/cgeo/geocaching/utils/OfflineMapUtils.java
+++ b/main/src/cgeo/geocaching/utils/OfflineMapUtils.java
@@ -20,6 +20,7 @@ import org.apache.commons.lang3.StringUtils;
 
 public class OfflineMapUtils {
 
+    public static final String MAPFILE_SUFFIX = ".map";
     public static final String INFOFILE_SUFFIX = "-cgeo.txt";
 
     private static final String PROP_PARSETYPE = "remote.parsetype";

--- a/tests/src-android/cgeo/geocaching/storage/ContentStorageTest.java
+++ b/tests/src-android/cgeo/geocaching/storage/ContentStorageTest.java
@@ -470,20 +470,21 @@ public class ContentStorageTest extends CGeoTestCase {
         final Folder testFolder = createTestFolder(type, "mimeType");
 
         performMimeTypeTests(testFolder,
-            new String[]{"txt", "jpg", "map"},
-            new String[]{"text/plain", "image/jpeg", "application/octet-stream"});
+            new String[]{"txt", "jpg", "jpeg", "map", "hprof", "gpx", null},
+            new String[]{"text/plain", "image/jpeg", "image/jpeg", "application/octet-stream", "application/octet-stream", "application/octet-stream", "application/octet-stream" });
     }
 
     private void performMimeTypeTests(final Folder testLocation, final String[] suffix, final String[] expectedMimeType) {
         final Map<String, String> mimeTypeMap = new HashMap<>();
         for (int i = 0; i < suffix.length; i++) {
-            final String filename = "test." + suffix[i];
+            final String filename = "test" + (suffix[i] == null ? "" : "." + suffix[i]);
             assertThat(ContentStorage.get().create(testLocation, filename)).isNotNull();
             mimeTypeMap.put(filename, expectedMimeType[i]);
         }
         final List<ContentStorage.FileInformation> files = ContentStorage.get().list(testLocation);
         assertThat(files.size()).isEqualTo(suffix.length);
         for (ContentStorage.FileInformation fi : files) {
+            assertThat(mimeTypeMap.containsKey(fi.name)).as("Unexpected File " + fi.name).isTrue();
             assertThat(fi.mimeType).as("For file " + fi.name).isEqualTo(mimeTypeMap.get(fi.name));
         }
     }

